test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl

```diff
@@ -0,0 +1,178 @@
+    if readPos + readSize > buffCap {
+      buffCap *= 2;
+      buffDom = {0..<buffCap};
+    }
+  } while newChars;
+
+  // if anything remains, process it
+  if readPos then revcomp(buff, readPos);
+}
+
+proc revcomp(seq, size) {
+  param chunkSize = linesPerChunk*cols;  // the size of the chunks to deal out
+
+  // compute how big the header is
+  var headerSize = 1;
+  while seq[headerSize - 1] != eol {
+    headerSize += 1;
+  }
+
+  // write out the header
+  stdoutBin.writeBinary(c_ptrTo(seq[0]), headerSize);
+
+  // set up the atomic variables we'll use to coordinate between tasks
+  var charsLeft, charsWritten: atomic int = size - (headerSize + 1);
+
+  // create a task per core
+  coforall tid in 0..<here.maxTaskPar {
+    var myBuff: [0..<chunkSize] uint(8);  // the task's buffer
+
+    var myStartChar = 0;
+    while getWork(myStartChar) {
+      const myChunkSize = min(chunkSize, myStartChar + 1),
+            lastLineChars = myStartChar % cols,
+            lastLineGaps = cols - (lastLineChars + 1);
+
+      var cursor = myStartChar + headerSize,
+          chunkLeft = myChunkSize,
+          chunkPos = 0;
+
+      if lastLineGaps == 0 {
+        revcomp(chunkPos, cursor, chunkLeft, myBuff, seq);
+      } else {
+        while chunkLeft >= cols {
+          revcomp(chunkPos, cursor, lastLineChars, myBuff, seq);
+          chunkPos += lastLineChars;
+          cursor -= lastLineChars + 1;
+
+          revcomp(chunkPos, cursor, lastLineGaps, myBuff, seq);
+          chunkPos += lastLineGaps;
+          cursor -= lastLineGaps;
+
+          myBuff[chunkPos] = eol;
+          chunkPos += 1;
+
+          chunkLeft -= cols;
+        }
+
+        if chunkLeft then
+          revcomp(chunkPos, cursor, lastLineChars + 1, myBuff, seq);
+      }
+
+      charsWritten.waitFor(myStartChar);
+      stdoutBin.writeBinary(c_ptrTo(myBuff[0]), myChunkSize);
+      charsWritten.write(myStartChar-myChunkSize);
+    }
+  }
+
+  // helper to atomically grab a chunk of work characterized by 'myStartChar'
+  inline proc getWork(ref myStartChar) {
+    myStartChar = charsLeft.read();
+    do {
+      if myStartChar < 0 then return false;
+    } while !charsLeft.compareExchange(myStartChar, myStartChar - chunkSize);
+
+    return true;
+  }
+}
+
+proc revcomp(in dstFront, in charAfter, spanLen, buff, seq) {
+  if spanLen%2 {
+    charAfter -= 1;
+    buff[dstFront] = cmpl[seq[charAfter]];
+    dstFront += 1;
+  }
+
+  for 2..spanLen by -2 {
+    charAfter -= 2;
+    const src = c_ptrTo(seq[charAfter]): c_ptr(uint(16)),
+          dst = c_ptrTo(buff[dstFront]): c_ptr(uint(16));
+    dst.deref() = pairCmpl[src.deref()];
+    dstFront += 2;
+  }
+}
+
+// see whether there's a sequence start ('>') in 'buff[low..#count]'
+proc findSeqStart(buff, low, count, ref ltOff) {
+  ltOff = max(int);
+  forall i in low..#count with (min reduce ltOff) {
+    if buff[i] == '>'.toByte() && i != 0 {
+      ltOff = i;
```

Comment:
Think this should be `ltOff = min(i, ltOff)`?

Reply:
Good catch, I'd been thinking this morning about whether I was just getting away with something here, and think you're right.  Unfortunately, adding the min takes the time down a notch further.

Reply:
Can you iterate in reverse or maybe do some sort of per task eurika?

```chpl
if == max(int) && buff[i] == '>'.toByte() && i != 0
```

or 

```chpl
forall i in low..#count by -1 ...
```

Not sure how optimized parallel negative strided range iterators are though..

Reply:
Hmm, and thinking some more -- it really surprises me that the `min` hurts performance. I was thinking there weren't that many instances of `>` so we wouldn't be calling min that often, but maybe that's wrong.

Reply:
Would it make a difference if ``ltOff`` were a local variable declared inside the function, rather than a ``ref`` argument?

Reply:
It doesn't seem to make much of a difference, which matches my intuition since most of the work on ltOff is being done on the task-local reduction variable, which will be a local variable anyway.

Reply:
The min() really is impactful on my Mac, though, resulting in 10% overhead...  Makes me want to cry because it was the change that took us from being competitive with the C version to still feeling a ways off from it (though we're obvs. still a ways off from it on the shootout box anyway).

Reply:
Forgot about this comment:

> Can you iterate in reverse or maybe do some sort of per task eurika?
> forall i in low..#count by -1 ...

Since foralls are order-independent, I'm not sure that "iterate in reverse" makes sense for them...